pushy/src/main/java/com/relayrides/pushy/apns/ApnsClientBuilder.java

```diff
@@ -0,0 +1,337 @@
+
+    /**
+     * <p>Sets the credentials for the client to be built using the data from the given PKCS#12 input stream. The
+     * PKCS#12 data <em>must</em> contain a single certificate/private key pair.</p>
+     *
+     * @param p12File an input stream to a PKCS#12-formatted file containing the certificate and private key to be used
+     * to identify the client to the APNs server
+     * @param p12Password the password to be used to decrypt the contents of the given PKCS#12 file; passwords may be
+     * blank (i.e. {@code ""}), but must not be {@code null}
+     *
+     * @throws SSLException if the given PKCS#12 file could not be loaded or if any other SSL-related problem arises
+     * when constructing the context
+     * @throws IOException if any IO problem occurs while attempting to read the given PKCS#12 file, or the PKCS#12 file
+     * could not be found
+     *
+     * @return a reference to this builder
+     *
+     * @since 0.8
+     */
+    public ApnsClientBuilder<T> setClientCredentials(final InputStream p12InputStream, final String p12Password) throws SSLException, IOException {
+        final X509Certificate x509Certificate;
+        final PrivateKey privateKey;
+
+        try {
+            final PrivateKeyEntry privateKeyEntry = P12Util.getPrivateKeyEntryFromP12InputStream(p12InputStream, p12Password);
+
+            final Certificate certificate = privateKeyEntry.getCertificate();
+
+            if (!(certificate instanceof X509Certificate)) {
+                throw new KeyStoreException("Found a certificate in the provided PKCS#12 file, but it was not an X.509 certificate.");
+            }
+
+            x509Certificate = (X509Certificate) certificate;
+            privateKey = privateKeyEntry.getPrivateKey();
+        } catch (final KeyStoreException | NoSuchAlgorithmException | UnrecoverableEntryException | CertificateException e) {
+            throw new SSLException(e);
+        }
+
+        return this.setClientCredentials(x509Certificate, privateKey, p12Password);
+    }
+
+    /**
+     * <p>Sets the credentials for the client to be built.</p>
+     *
+     * @param clientCertificate the certificate to be used to identify the client to the APNs server
+     * @param privateKey the private key for the client certificate
+     * @param privateKeyPassword the password to be used to decrypt the private key; may be {@code null} if the private
+     * key does not require a password
+     *
+     * @return a reference to this builder
+     *
+     * @since 0.8
+     */
+    public ApnsClientBuilder<T> setClientCredentials(final X509Certificate clientCertificate, final PrivateKey privateKey, final String privateKeyPassword) {
+        this.clientCertificate = clientCertificate;
+        this.privateKey = privateKey;
+        this.privateKeyPassword = privateKeyPassword;
+
+        return this;
+    }
+
+    protected ApnsClientBuilder<T> setTrustedServerCertificate(final File trustedServerCertificatePemFile) {
```

Comment:
I'd like to publicize this for certificate pinning, but don't yet understand Netty's trust manager construction well enough to be confident in the approach. I intend to do some more research here.